doc/Hash/Policy-based-hash.md

```diff
@@ -0,0 +1,215 @@
+# Policy-based Hash
+
+Title       | Policy-based Hash
+------------|----------------
+Authors     | Nvidia
+Status      | In review
+Type        | Standards track
+Created     | 04/29/2020
+SAI-Version | 1.6
+
+Sai defines two metadata values for ECMP and LAG hash.
+The fields that are used for their calculation are defined at the switch level with attributes `SAI_SWITCH_ATTR_ECMP_HASH` and `SAI_SWITCH_ATTR_LAG_HASH`.
+There are also hash objects that can be assigned for some basic packet types.
+
+But if we would like to cover some use cases that are not defined in SAI, we need to be more flexible in a way that won't require changing the API.
+These use cases will be covered by a concept introduced here called policy-based hashing.
+
+```
+    /** Set custom LAG hash object ID */
+    SAI_ACL_ACTION_TYPE_SET_LAG_HASH_ID,
+
+    /** Set custom ECMP hash object ID */
+    SAI_ACL_ACTION_TYPE_SET_ECMP_HASH_ID,
+
+} sai_acl_action_type_t;
+```
+
+The two action types are not conflicting between one another, but the same action type used more than once for the parralel execution is.
+
+It utilizes the ACL infrastructure with the new actions to set custom hash object for a special ECMP or LAG treatment.
+
+Also, the hash object itself is extended to be able to specify the portions of the fields that we are interested in, like, for example, only some bytes from IPv6 address.
+
+```
+    /**
+     * @brief Hash fine-grained field list
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_FINE_GRAINED_HASH_FIELD
+     * @default empty
+     */
+    SAI_HASH_ATTR_FINE_GRAINED_HASH_FIELD_LIST,
+```
+
+Optional ordering of the fields is available as well.
+
+```
+    /**
+     * @brief Optional field ordering.
+     *
+     * Specifies in which order the fields are hashed,
+     * and defines in which fields should be associative for CRC with the same sequence ID.
+     * If not provided, it's up to SAI driver to choose.
```

Comment:
Hi, can we please add in the doc why "sequence id" is needed for the hashing fields?  I recall this is needed so that bidirectional flows can get to the same hash.  What does it mean by "defines in which fields should be associative for CRC with the same sequence ID"?

Reply:
Added a description. Also posting it here as a reply:
```
In the case where the traffic is represented by two different packet types for inbound and outbound direction, 
setting the hash order is neccessary for non-associative hash algorithms like CRC, so that the resulting hash 
value will be the same. For example, for the bare metal machine sending outbound IP packet to the ToR, the hash 
field array could look like this: {SRC_IP:0, DST_IP:1, SRC_PORT:2, DST_PORT:3, IP_TYPE:4}, and for the inbound, getting 
VxLAN packet from the T1 to the same ToR, the hash field array will look like this: {INNER_DST_IP:0, INNER_SRC_IP:1, 
INNER_DST_PORT:2, INNER_SRC_PORT:3, INNER_IP_TYPE:4}. Here the hash associativity is achieved by swapping 
source and destination sequence IDs. The other way to achieve this would be to assign the same shared sequence ID 
value for IP addresses, and another for L4 ports.
```

Reply:
@marian-pritsak Assigning same sequence Id translates to what order ? 
For ex : src and dest ip having sequence Id will the order be {src-ip, dest-ip } or {dest-ip, src-ip} How the order is guarantee from SDK perspective ?